### ISSUE TITLE
Prose wrap options for Korean

### ISSUE THREAD

**Issue by Rokt33r** [prettier/prettier#6516]
Korean language should be formatted like English. Sadly, prettier could not recognize Korean.
 
**Prettier 1.18.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEgAZsDqrgf7sLa1AKQxy2AMdQAS6AjzYANjgC50DtgqBOAvPQJQA0xg5JqDL6YIg2g0OpuBAMcAto0mIAlOAAcArgCMANgEswxCADNiAaQgAnOAEMA3OIDymgNoBmALpNAAuPFAm82ARUcA7Q88Aa48XSAMHrLFAREnACobAAN7ABdHiQAjxwBSm4kBXmsAVNexAF3HAE6aAOgAdKGzAHEHAXqnAHnHAHQ7Y3MAXVcAfTuJAGoHkiMAFFsADmsCEwA9xzESUtkci0qcq2vriZuJAH3HAD57kqMLABy7swEDevtjABwnAAkHMQB+JzY8OrsAfmuJAPc7Af6bnKIyQFhAISRhFaABnZFB9HR0IAHcABTeEZ5Q+gAbhBFAATK4gWQ6fRgADWcBgAGVJLDFFAAObIGA6aRwa4ACxgAFt5AB1AmKeCPVFgOBI-5UxRAqkAT2Q4Eez2u6MecB0MG+MIxxP0yDU+nkfOuACtHgAPABCMPhiKR+mJcAAMui4OLJdKQHL5Uj0Rj5HAAIrSCDwfVS-EgVE6Pk6DminRwsFfKCQyQ6dEwMngmAE5AADgADNd-RA+WSYZIOZLPvpWdyQABHG3wIW3AEgfSPAC0UDgcDBFchemzij0Qv0IrFSAlDuufOJimxuMdjzNFuttr1LYNjpg+lkwbBoeQACZrjj9IolJiAMIQYmijlQaB667SPkAFQnANbfIAvuegA)

```sh
# Options (if any):
--prose-wrap=always
```

**Input:**
```
대한민국(듣기 (도움말·정보), 大韓民國, 영어: Republic of Korea; ROK[3])은 동아시아의 한반도 중남부에 있는 공화국이다.

서쪽으로는 서해를 사이에 두고 중화인민공화국이, 동쪽으로는 동해를 사이에 두고 일본이 있으며
북쪽으로는 조선민주주의인민공화국과 맞닿아 있다.
```

**Output:**
```
대한민국(듣기 (도움말·정보), 大韓民國, 영어: Republic of Korea; ROK[3])은 동아시
아의 한반도 중남부에 있는 공화국이다.

서쪽으로는 서해를 사이에 두고 중화인민공화국이, 동쪽으로는 동해를 사이에 두고 일
본이 있으며북쪽으로는 조선민주주의인민공화국과 맞닿아 있다.
```

The example shows three problems.

- `"동아시아"` in the first paragraph is a single word. But Prettier splits it into `"동아시\n아"`.
- `"일본"` in the second paragraph is also a single word. It should not be split into `"일\n본"`
- `"있으며북쪽으로는"` in the second paragraph should be two words. `"있으며 북쪽으로는"`

**Expected behavior:**

```
대한민국(듣기 (도움말·정보), 大韓民國, 영어: Republic of Korea; ROK[3])은 동아시아의 한반도
중남부에 있는 공화국이다.

서쪽으로는 서해를 사이에 두고 중화인민공화국이, 동쪽으로는 동해를 사이에 두고 일본이 있으며 북쪽으로는
조선민주주의인민공화국과 맞닿아 있다.
```

**Comment by thorn0** [comment]
> Korean is unusual in that words in modern hangul text are normally separated by spaces, but the writing system allows content authors to choose one of two ways for that text to be wrapped.
> Syllable-based wrapping is common, especially in fully justified text (which is more common in CJK writing systems than in Western ones), but paragraphs with a ragged right edge will often wrap whole words. However, the choice is motivated by author preference, rather than any hard and fast rule.

-- from https://www.w3.org/International/articles/typography/linebreak

Sounds like the current behavior is valid too, so it's not a bug. Also my browser (Chrome) breaks lines the same way, e.g. on [Korean Wikipedia](https://ko.wikipedia.org/wiki/%EB%8C%80%ED%95%9C%EB%AF%BC%EA%B5%AD).

**Comment by thorn0** [comment]
> "있으며북쪽으로는" in the second paragraph should be two words. "있으며 북쪽으로는"

This actually looks like a bug.

**Comment by thorn0** [comment]
It can't be fixed, can it? When re-wrapping text (`--prose-wrap` ≠ `preserve`), Prettier has no way to detect whether a line break in the input was inserted in the middle of a word to wrap it or between words, which is necessary to choose what that line break should be replaced with: with the empty string or with a space. This, unfortunately, means it's impossible to implement `--prose-wrap` properly for Korean, and the default value `preserve` should be always used for this language.

I'm leaving this open because I actually know nothing about CJK. Hopefully, somebody who knows more will chime in and prove my conclusion wrong.

@ikatyang I'm interested to know your opinion on this.

**Comment by ikatyang** [comment]
We should probably force `--prose-wrap preserve` for Korean since they use conventional spaces (ref: #5028), it's not possible for us to know whether we should break or join the line just like you mentioned above.

**Comment by 18kimn** [comment]
Any updates on this? I feel that simply treating Korean words like English text for word-wrap would be a better behavior. The current system (either turning off Prettier or using `--prose-wrap preserve`) to me means that Prettier just does not support Korean markdown, since separate words across multiple lines get joined together by Prettier. It can get really bad when I edit and re-edit a document and these conjoined words pile up. Prettier is great to me in every other aspect, so I'm sad about this lack of support.

The justification for syllabus-based wrapping is not a set-in-stone rule, as the quote pulled by @thorn0 suggests, so I think departing from it to avoid broken behavior would be okay. I can make a PR, but am not sure where to look for this behavior (any pointers would be appreciated)

thank yoou!

**Comment by tats-u** [comment]
I am trying to fix this issue in https://github.com/prettier/prettier/pull/11597.
Could you try this playground <https://deploy-preview-11597--prettier.netlify.app/playground/> and tell me your review?

```
한국어와 English를 섞어 보았습니다. 한국어와 English를 섞어 보았습니다.
한국어와
English를
섞어
보았습니다.
한국어와
English를
섞어
보았습니다.
한국어와 English를 섞어 보았습니다. 한국어와 English를 섞어 보았습니다.
한국어와
English를
섞어
보았습니다.
한국어와
English를
섞어
보았습니다.
한국어와 English를 섞어 보았습니다. 한국어와 English를 섞어 보았습니다.

NVIDIA의 RTX3000 시리즈는 삼성전자와 TSMC에 의해 제조된다.
NVIDIA의 RTX3000 시리즈는 삼성전자와 TSMC에 의해 제조된다.
NVIDIA의
RTX3000
시리즈는
삼성전자와
TSMC에
의해
제조된다.
NVIDIA의 RTX3000 시리즈는 삼성전자와 TSMC에 의해 제조된다.
NVIDIA의 RTX3000 시리즈는 삼성전자와 TSMC에 의해 제조된다.

대한민국(듣기 (도움말·정보), 大韓民國, 영어: Republic of Korea; ROK[3])은 동아시아의 한반도 중남부에 있는 공화국이다.

서쪽으로는 서해를 사이에 두고 중화인민공화국이, 동쪽으로는 동해를 사이에 두고 일본이 있으며
북쪽으로는 조선민주주의인민공화국과 맞닿아 있다.
```

↓ `proseWrap: "always"`

```
한국어와 English를 섞어 보았습니다. 한국어와 English를 섞어 보았습니다. 한국어와
English를 섞어 보았습니다. 한국어와 English를 섞어 보았습니다. 한국어와
English를 섞어 보았습니다. 한국어와 English를 섞어 보았습니다. 한국어와
English를 섞어 보았습니다. 한국어와 English를 섞어 보았습니다. 한국어와
English를 섞어 보았습니다. 한국어와 English를 섞어 보았습니다.

NVIDIA의 RTX3000 시리즈는 삼성전자와 TSMC에 의해 제조된다. NVIDIA의 RTX3000
시리즈는 삼성전자와 TSMC에 의해 제조된다. NVIDIA의 RTX3000 시리즈는 삼성전자와
TSMC에 의해 제조된다. NVIDIA의 RTX3000 시리즈는 삼성전자와 TSMC에 의해 제조된다.
NVIDIA의 RTX3000 시리즈는 삼성전자와 TSMC에 의해 제조된다.

대한민국(듣기 (도움말·정보), 大韓民國, 영어: Republic of Korea; ROK[3])은
동아시아의 한반도 중남부에 있는 공화국이다.

서쪽으로는 서해를 사이에 두고 중화인민공화국이, 동쪽으로는 동해를 사이에 두고
일본이 있으며 북쪽으로는 조선민주주의인민공화국과 맞닿아 있다.
```

Also I have another question: how about newline between hangul and hanja? Do you expect

```
선거 결과 尹錫悅가 대통령으로 취임했다.
```

from

```
선거 결과
尹錫悅가 대통령으로 취임했다.
```

? Do you allow to break line between `悅` and `가`?

**Comment by olimot** [comment]
@tats-u

All other cases exept the last one looks good. Thank you!

My problem was that prettier inserts a new whitespace after prettier breaks a line, which did not exist before formatting. 

In Korean language, it allows to break line between 悅 and 씨 and actually prefered to adding a whitspace between them. But the problem is when a markdown document is about what I'm talking about right above.

```
예를 들어 "尹錫悅씨가"의 경우,  "尹錫悅 씨가"와 같이 띄어 쓰는 것을 원칙으로 한다.
```

↓ prettier (proseWrap: "always")

```
예를 들어 "尹錫悅
씨가"의 경우,
"尹錫悅 씨가"와
같이 띄어 쓰는
것을 원칙으로
한다.
```

↓ any other markdown viewer (below is in github markdown viewer)

```
예를 들어 "尹錫悅 씨가"의 경우,  "尹錫悅 씨가"와 같이 띄어 쓰는 것을 원칙으로 한다.
```

It changes its meaning after formatting. The example before prettier means:

In case of "尹錫悅씨가", there should be a whitespace like "尹錫悅 씨가" in convention of writing Korean.

But the last one in github markdown viewer, the first "尹錫悅씨가" becomes "尹錫悅 씨가", so it lost what i wanted to say originally)

**Comment by tats-u** [comment]
So can Prettier treat the combination of Chinese & Korean as that of Korean & Korean?

- `悅\n씨` → `悅 씨`
- ` 悅|씨` (`|` means the line length limit)l → `\n悅씨` (Don't break)

**Comment by olimot** [comment]
Yes you're right. Thank you again and I appreciate your work

**Comment by tats-u** [comment]
@olimot Thank you for your review. Could you review the patch note as well? I regret I know little about Korean.

https://github.com/prettier/prettier/pull/11597/commits/24679500cb1078774a53ffd530e882f63a15e41b

https://github.com/prettier/prettier/blob/24679500cb1078774a53ffd530e882f63a15e41b/changelog_unreleased/markdown/11597.md?plain=1#L114-L155

@Rokt33r @18kimn You are welcome to review my patch, too. I would be grad if you help me.

**Comment by olimot** [comment]
The example in the note is perfectly fit to what I wanted to say. Thank you!

**Comment by thorn0** [comment]
Fixed by #11597

**Comment by tats-u** [comment]
Fixed in 3.0.0-alpha.3?

```
PS> npx prettier@3.0.0-alpha.2  --prose-wrap=always  .\test.md
# 第 1 章 あいうえお

Prettier で Markdown をフォーマットしましょう。

대한민국(듣기 (도움말·정보), 大韓民國, 영어: Republic of Korea; ROK[3])은 동아시
아의 한반도 중남부에 있는 공화국이다.

서쪽으로는 서해를 사이에 두고 중화인민공화국이, 동쪽으로는 동해를 사이에 두고 일
본이 있으며북쪽으로는 조선민주주의인민공화국과 맞닿아 있다.
PS> npx prettier@3.0.0-alpha.3  --prose-wrap=always  .\test.md
# 第1章 あいうえお

PrettierでMarkdownをフォーマットしましょう。

대한민국(듣기 (도움말·정보), 大韓民國, 영어: Republic of Korea; ROK[3])은
동아시아의 한반도 중남부에 있는 공화국이다.

서쪽으로는 서해를 사이에 두고 중화인민공화국이, 동쪽으로는 동해를 사이에 두고
일본이 있으며 북쪽으로는 조선민주주의인민공화국과 맞닿아 있다.
```
